capreolus/sampler/__init__.py

```diff
@@ -0,0 +1,97 @@
+    """
+
+    def __init__(self, search_run, benchmark, extractor):
+        self.search_run = search_run
+        self.benchmark = benchmark
+        self.extractor = extractor
+
+    def __iter__(self):
+        """
+        Returns: Triplets of the form (query_feature, posdoc_feature, negdoc_feature)
+        """
+        extractor = self.extractor
+        benchmark = self.benchmark
+        search_run = self.search_run
+
+        train_qids = benchmark.folds[benchmark.cfg["fold"]]["train_qids"]
+        is_rundocsonly = benchmark.cfg["rundocsonly"]
+        if is_rundocsonly:
+            # Create qrels from only those docs that showed up in the search results
+            qrels = {
+                qid: {docid: label for docid, label in benchmark.qrels[qid].items() if docid in search_run[qid]}
+                for qid in train_qids
+            }
+        else:
+            qrels = benchmark.qrels
+
+        qid_to_reldocs = {qid: [docid for docid, label in qrels[qid].items() if label > 0] for qid in qrels}
+        qid_to_negdocs = {qid: [docid for docid, label in qrels[qid].items() if label <= 0] for qid in qrels}
+
+        # Remove the qids that does not have relevant/irrelevant docs associated with it
+        valid_qids = [qid for qid in train_qids if qid_to_reldocs.get(qid) and qid_to_negdocs.get(qid)]
+
+        # Convert each query and doc id to the corresponding feature/embedding and yield
+        def genf():
+            while True:
+                # random.shuffle(valid_qids)
+
+                for qid in valid_qids:
+                    posdocid = random.choice(qid_to_reldocs[qid])
+                    negdocid = random.choice(qid_to_negdocs[qid])
+
+                    try:
+                        query_feature, posdoc_feature, negdoc_feature = extractor.id2vec(qid, posdocid, negdocid)
+                        yield {"query": query_feature, "posdoc": posdoc_feature, "negdoc": negdoc_feature}
+                    # TODO: Replace below catch-all exception with MissingDocError
+                    except Exception:
+                        # The extractor should emit the above error
+                        logger.warning("got none features: qid=%s posid=%s negid=%s", qid, posdocid, negdocid)
+                        raise
+
+        return genf()
+
+
+class PredDataset(torch.utils.data.IterableDataset):
+    """
+    Creates a Dataset for evaluation (test) data to be used with a pytorch DataLoader
+    """
+
+    def __init__(self, search_run, benchmark, extractor):
+        self.search_run = search_run
+        self.benchmark = benchmark
+        self.extractor = extractor
+
+    def __iter__(self):
+        """
+        Returns: Tuples of the form (query_feature, posdoc_feature)
+        """
+        benchmark = self.benchmark
+        extractor = self.extractor
+        search_run = self.search_run
+        test_qids = benchmark.folds[benchmark.cfg["fold"]]["predict"]["test"]
```

Comment:
I guess we still need a Dataset for dev? or maybe we can put the train/dev/test as an __init__ argument and let them share the genf() code

Reply:
I think we should decouple this more from the train/dev/test idea. We can pass in qrels and the valid `(qid, docid)` pairs to be considered, so that the Dataset doesn't have to know anything about the benchmark, searcher, or folds. Then we can have one dataset that samples following some strategy (the training one currently) and one dataset that asks for one prediction for each training instance (the pred one). I'll do this when I integrate with the trainer